structure_tests/drivers/docker_driver.go

```diff
@@ -0,0 +1,290 @@
+	}
+	stdout, stderr, exitCode := d.exec(t, env, fullCommand)
+
+	if stdout != "" {
+		t.Logf("stdout: %s", stdout)
+	}
+	if stderr != "" {
+		t.Logf("stderr: %s", stderr)
+	}
+	//reset image for next test
+	d.currentImage = d.originalImage
+	return stdout, stderr, exitCode
+}
+
+func (d *DockerDriver) processEnvVars(t *testing.T, vars []unversioned.EnvVar) []string {
+	if len(vars) == 0 {
+		return nil
+	}
+
+	image, err := d.cli.InspectImage(d.currentImage)
+	if err != nil {
+		t.Errorf("Error when inspecting image: %s", err.Error())
+		return nil
+	}
+
+	// convert env to map for easier processing
+	imageEnv := make(map[string]string)
+	for _, varPair := range image.Config.Env {
+		pair := strings.Split(varPair, "=")
+		imageEnv[pair[0]] = pair[1]
+	}
+
+	before := regexp.MustCompile(".*\\$(.*?):")
+	after := regexp.MustCompile(".*:\\$(.*)")
+
+	env := []string{}
+	for _, envVar := range vars {
+		currentVar := ""
+		if match := before.FindStringSubmatch(envVar.Value); match != nil {
+			// first entry is the leftmost substring: second entry is the first group
+			currentVar = match[1]
+		}
+		if match := after.FindStringSubmatch(envVar.Value); match != nil {
+			currentVar = match[1]
+		}
+		if currentVar != "" {
+			if val, ok := imageEnv[currentVar]; ok {
+				env = append(env, envVar.Key+"="+strings.Replace(envVar.Value, "$"+currentVar, val, -1))
+			} else {
+				t.Errorf("Variable %s not found in image env! Check test config.", currentVar)
+				return nil
+			}
+		} else {
+			env = append(env, envVar.Key+"="+envVar.Value)
+		}
+	}
+	return env
+}
+
+// copies a file from a docker container to the local fs, and returns its path
+// caller is responsible for removing this file when finished
+func (d *DockerDriver) retrieveFile(t *testing.T, path string) (string, error) {
+	// this contains a placeholder command which does not get run, since
+	// the client doesn't allow creating a container without a command.
+	container, err := d.cli.CreateContainer(docker.CreateContainerOptions{
+		Config: &docker.Config{
+			Image:        d.currentImage,
+			Cmd:          []string{"NOOP_COMMAND_DO_NOT_RUN"},
+			AttachStdout: true,
+			AttachStderr: true,
+		},
+		HostConfig:       nil,
+		NetworkingConfig: nil,
+	})
+	if err != nil {
+		t.Errorf("Error creating container: %s", err.Error())
+		return "", err
+	}
+
+	tmpFile, err := ioutil.TempFile("", "structure_test")
+	if err != nil {
+		t.Errorf("Error when creating temp file: %s", err.Error())
+		return "", err
+	}
+	stream := bufio.NewWriter(tmpFile)
+
+	err = d.cli.DownloadFromContainer(container.ID, docker.DownloadFromContainerOptions{
+		OutputStream: stream,
+		Path:         path,
+	})
+	if err != nil {
+		t.Errorf("Error when downloading file from container: %s", err.Error())
+		return "", err
+	}
+	stream.Flush()
+	tmpFile.Close()
+	return tmpFile.Name(), nil
+}
+
+func (d *DockerDriver) StatFile(t *testing.T, path string) (os.FileInfo, error) {
```

Comment:
I think we probably want to reconsider this part of the interface a bit. Statting the file this way won't actually be accurate, since you lose things like atime/mtime and permissions the way this driver copies the files out of the image.

It also doesn't look like we actually use the stat results anywhere right now, other than as a basic existence check.

Reply:
hmm, do we lose the permissions? looks like at the very least, fileMode and mTime are preserved and returned back from the Docker API: https://docs.docker.com/engine/api/v1.23/#2-endpoints

we use the stat results to check permissions/mode in the FileExistence tests.